### ISSUE TITLE
Huginn can't run on Ubuntu v22 because of Ruby/OpenSSL v3 incompatibilities (and Ubuntu v20 maintenance updates cease in 2025)

### ISSUE THREAD

**Issue by ohingardail** [huginn/huginn#3149]
According to the [Huginn installation instructions](https://github.com/huginn/huginn/blob/master/doc/manual/installation.md), Huginn requires Ruby v.2.6.9. Ruby v2.6.9 requires OpenSSL v1. OpenSSL v1 has been replaced by OpenSSL v3 on Ubuntu v22. The [Huginn requirements document](https://github.com/huginn/huginn/blob/master/doc/manual/requirements.md) states that only Ubuntus v18, v16 and v14 are supported.

In fact, Huginn can run on Ubuntu v20 (LTS) and v21 (interim; now EOL) without too much difficulty; you need to install OS packages libgdbm-compat-dev (to allow Ruby v2.6.9 compilation using old database header files) and runit-systemd (as with Ubuntu v18), and not install OS package python-docutils (which appears to have disappeared). Both of these Ubuntu versions include OpenSSL v1.

OpenSSL on Ubuntu v20 (abridged output)
```
> lsb_release -a
Description:	Ubuntu 20.04.4 LTS
> openssl version
OpenSSL 1.1.1f  31 Mar 2020
```

OpenSSL on Ubuntu v22 (abridged output)
```
> lsb_release -a
Description:	Ubuntu 22.04 LTS
> openssl version
OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022)
```

Upgrading a running Huginn server from v20/v21 to v22 breaks the Ruby v2.6.9 installation (abridged output for concision):

```
> lsb_release -a
Description:	Ubuntu 20.04.4 LTS                             # Initial Ubuntu version
> ruby -v
ruby 2.6.9p207 (2021-11-24 revision 67954) [x86_64-linux]      # Ruby version
> openssl version
OpenSSL 1.1.1f  31 Mar 2020                                    # OpenSSL version

> cd ~huginn/huginn
> sudo sudo bundle exec rake production:stop                   # Shut down Huginn in the usual way (no errors)

> sudo do-release-upgrade                                      # The actual Ubuntu version update command; lots of abridged output here (but no errors)

> lsb_release -a
Description:	Ubuntu 22.04 LTS                               # Final Ubuntu version
> ruby -v
ruby 2.6.9p207 (2021-11-24 revision 67954) [x86_64-linux]      # Ruby version
> openssl version
OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022) # OpenSSL version

> sudo bundle exec rake production:start --trace
rake aborted!
LoadError: libssl.so.1.1: cannot open shared object file: No such file or directory - /usr/local/lib/ruby/2.6.0/x86_64-linux/openssl.so
/usr/local/lib/ruby/2.6.0/openssl.rb:13:in `require'
/usr/local/lib/ruby/2.6.0/openssl.rb:13:in `<top (required)>'
```

You cannot recompile Ruby v2.6.9 using OpenSSL v3; it specifies that it needs OpenSSL v1. This is what you see if you attempt to compile Ruby v2.6.9 on an Ubuntu v22 box (abridged output):
```
>make -j`nproc`
:
make[2]: Entering directory '/tmp/ruby/ruby-2.6.9/ext/openssl'
compiling openssl_missing.c
compiling ossl.c
ossl.c: In function ‘ossl_clear_error’:
ossl.c:311:9: error: ‘ERR_get_error_line_data’ is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  311 |         while ((e = ERR_get_error_line_data(&file, &line, &data, &flags))) {
      |         ^~~~~
In file included from ossl.h:21,
                 from ossl.c:10:
/usr/include/openssl/err.h:413:15: note: declared here
  413 | unsigned long ERR_get_error_line_data(const char **file, int *line,
      |               ^~~~~~~~~~~~~~~~~~~~~~~
ossl.c: At top level:
cc1: note: unrecognized command-line option ‘-Wno-self-assign’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-parentheses-equality’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-constant-logical-operand’ may have been intended to silence earlier diagnostics
cc1: some warnings being treated as errors
make[2]: *** [Makefile:313: ossl.o] Error 1
make[2]: Leaving directory '/tmp/ruby/ruby-2.6.9/ext/openssl'
make[1]: *** [exts.mk:243: ext/openssl/all] Error 2
make[1]: Leaving directory '/tmp/ruby/ruby-2.6.9'
make: *** [uncommon.mk:286: build-ext] Error 2
```

The compilation of the Huginn-unsupported (ie, it may work, but it's not documented as such) Ruby version v2.7.6 includes this somewhat more useful error (abridged output):

```
> lsb_release -a
Description:	Ubuntu 22.04 LTS
>make -j`nproc`
*** Following extensions are not compiled:
openssl:
	Could not be configured. It will not be installed.
	/tmp/ruby/ruby-2.7.6/ext/openssl/extconf.rb:111: OpenSSL >= 1.0.1, < 3.0.0 or LibreSSL >= 2.5.0 is required
	Check ext/openssl/mkmf.log for more details.
*** Fix the problems, then remove these directories and try again if you want.
```
The referenced [mkmf.log](https://github.com/huginn/huginn/files/9212443/mkmf.log) file only says the same thing at greater length (such as including gcc commandlines and checked program details), including:
```
checking for OpenSSL version >= 1.0.1 and < 3.0.0... -------------------- no
/tmp/ruby/ruby-2.7.6/ext/openssl/extconf.rb:111: OpenSSL >= 1.0.1, < 3.0.0 or LibreSSL >= 2.5.0 is required
```

There may be versions of Ruby that both support OpenSSL v3 and Huginn, but I don't know what these would be, or how much effort it would take to get it working reliably. It might also be possible to hack Ubuntu v22 to use OpenSSL v1 or LibreSSL, but my brief forays into exploring these options suggest a world of compatibility problems, broken system components, potential security flaws and fragile upgrade paths. 

Since Ubuntu v21 was an interim release which has been EOLed, this suggests the latest version of Ubuntu upon which Huginn can currently be made to work reliably is v20, for which there will be [maintenance updates until 2025 and security updates until 2030](https://ubuntu.com/about/release-cycle). The [Huginn requirements document](https://github.com/huginn/huginn/blob/master/doc/manual/requirements.md) specify Ubuntu v18 as the latest supported version, which will receive [maintenance updates until 2023 and security updates until 2028](https://ubuntu.com/about/release-cycle).

My assumption, provided without evidence but based on previous experience with relying on old versions of software components for too long, is that even if the issues with Ruby/OpenSSL compatibility on Ubuntu v22 were resolved, more things will break in unexpected and exotic ways as time rolls by, with each new customised workaround potentially adding problems later down the line. Eventually, existing Huginn users will be marooned in an old and unsupported version of Ubuntu (possibly held together with undocumented workarounds), and potential new users won't be able to obtain the Ubuntu OS pre-requisites to even install a working Huginn to start with.

**Comment by ohingardail** [comment]
I could update the Huginn Ubuntu [installation](https://github.com/huginn/huginn/blob/master/doc/manual/installation.md) and requirements pages to include Ubuntu v20, of I'm given permissions so to do. It would put off the evil hour until 2025, at least.

**Comment by rndph** [comment]
Well Ruby 2.6 is EOL even by itself. So I feel that should be upgraded first: https://github.com/huginn/huginn/pull/3055

**Comment by virtadpt** [comment]
Ruby v2.6.3p62 works as expected.  Installed with RVM.

**Comment by virtadpt** [comment]
Well, I'm an idiot.  I upgraded my box to Ubuntu 22.04.1 LTS tonight to test this, and everything blew up in my face.

**Comment by virtadpt** [comment]
I think I have a partial fix.
The first step is to get the rvm-installed Ruby environment linked against a compatible OpenSSL library.  Here are the steps I used for that:

```
rvm remove <your installed version here>
rvm cleanup all
rvm get head
rvm pkg install openssl
rvm install 2.6 -C --with-openssl-dir=$HOME/.rvm/usr
```
However, you'll run into some additional problems once you've solved that one.

**Comment by virtadpt** [comment]
The biggest problem you will run into is that any time you try to do anything that accesses the database (e.g., `rake db:migrate`) it'll fail with a segfault, like this one:

```
...
/home/drwho/.rvm/gems/ruby-2.6.0/gems/mysql2-0.5.4/lib/mysql2/client.rb:95: [BUG
] Segmentation fault at 0x0000000000000000
ruby 2.6.0p0 (2018-12-25 revision 66547) [x86_64-linux]
...
/home/drwho/.rvm/gems/ruby-2.6.0/gems/mysql2-0.5.4/lib/mysql2/client.rb:95:in `c
onnect'

-- Machine register context ------------------------------------------------
 RIP: 0x00007f2857bd4ed0 RBP: 0x00007f2858f6aaf0 RSP: 0x00007ffe9fa91aa0
 RAX: 0x001f0ffff715e3e9 RBX: 0x00007ffe9fa92010 RCX: 0x00000000acde9668
 RDX: 0x0000000000000078 RDI: 0x00007f2858f6aaf0 RSI: 0x0000000000000400
  R8: 0x0000000000000001  R9: 0x000000005e18771f R10: 0x0000000047795fa1
 R11: 0x000000005aafc017 R12: 0x000055db608f8730 R13: 0x00007ffe9fa91b28
 R14: 0x00007ffe9fa91f50 R15: 0x000000000000001e EFL: 0x0000000000010202

-- C level backtrace information -------------------------------------------
...
7f2856803000-7f2856804000 r--p 00000000 fc:01 1704025                    /usr/lib/x86_64-linux-gnu/libmariadb3/plugin/caching_sha2_password.so
...
```
I'm still working out where in the call chain this is happening.  I strongly suspect that it has to do with the Ubuntu v22.04 dropping all support for OpenSSL v1.0.

**Comment by virtadpt** [comment]
A seemingly fruitful search term: "ubuntu 22.04.1 mysql2 gem segmentation fault"

**Comment by virtadpt** [comment]
I really hate to say it, but it seems running Huginn on Ubuntu 22.x is a no-go.  The libssl mismatch and pretty much any version of Ruby (the farthest back I tried was v2.3.0 (all tests run with RVM, which I normally use)) just don't work - specifically, the mysql2 gem (I tried everything from v0.4.0 to v0.5.4) breaks on the `connect()` method with a segfault.  I wound up building a brand new Ubuntu 20.04 LTS server and migrating all of my stuff over to it.

If setting up a 20.04 LTS server is an option for you, that might be the only thing that'll work.

**Comment by ohingardail** [comment]
Well, I'm sort-of glad that the sense of having a plateful of cold spaghetti being thrown over oneself, which is trying to get Huginn to work on Ubuntu v22, isn't just me! It took hours to scrub the stains out of the carpet...

Until Huginn is updated, I think it would be prudent to update the [installation docs](https://github.com/huginn/huginn/tree/master/doc/manual) to show that Huginn can work on Ubuntu v20 (with a few [tweaks](https://github.com/huginn/huginn/issues/3149#issue-1321143099)), but attempting the same on v22 is not supported (and v21 is EOL). People trying out Huginn for the first time might assume they should use the latest version of Ubuntu, and that won't work.

**Comment by dohlin** [comment]
Also glad to hear this isn't just.  Updated my install as well and thankfully I snap'ed the vm beforehand.  Hit my head against the wall for a while on it before saying screw it and rolling back.

**Comment by virtadpt** [comment]
I ran into a somewhat similar problem after upgrading one of my remote job worker nodes (Arch Linux btw, as of 1 May 2023).  The package openssl-1.0 is necessary.  Two steps were involved:

* `export PKG_CONFIG_PATH=/usr/lib/openssl-1.0/pkgconfig`
* `rvm install 2.7 --with-openssl-dir=/usr/lib/openssl-1.0`

I tried Ruby v3.0 but it didn't work reliably; going back to Ruby v2.7 is working at this time.

If I can pull together a more up-to-date Ubuntu LTS install I'll give it a try there.
